packages/lisk-transaction-pool/src/transaction_pool.ts

```diff
@@ -40,62 +40,136 @@ interface Queues {
 	readonly [queue: string]: Queue;
 }
 
-interface ProcessTransactionsResponse {
-	readonly errors: ReadonlyArray<Error>;
-	readonly invalidTransactions: ReadonlyArray<Transaction>;
-	readonly validTransactions: ReadonlyArray<Transaction>;
-}
-
-type processTransactions = (
-	transactions: ReadonlyArray<Transaction>,
-) => ProcessTransactionsResponse;
-
 export class TransactionPool {
-	private applyTransactions: processTransactions;
-	private queues: Queues;
-	private validateTransactions: processTransactions;
-	private verifyTransactions: processTransactions;
-
-	public constructor(
-		validateTransactions: processTransactions,
-		verifyTransactions: processTransactions,
-		applyTransactions: processTransactions,
-	) {
-		this.queues = {
+	// tslint:disable-next-line variable-name
+	private readonly _queues: Queues;
+
+	public constructor() {
+		this._queues = {
 			received: new Queue(),
 			validated: new Queue(),
 			verified: new Queue(),
 			pending: new Queue(),
 			ready: new Queue(),
 		};
-		this.validateTransactions = validateTransactions;
-		this.verifyTransactions = verifyTransactions;
-		this.applyTransactions = applyTransactions;
 	}
 
-	public addTransactions(transactions: ReadonlyArray<Transaction>): void {}
+	public addTransactions(transactions: ReadonlyArray<Transaction>): void {
+		transactions.forEach((transaction: Transaction) => {
+			if (this.existsInTransactionPool(transaction)) {
+				this._queues.received.enqueueOne(transaction);
+			}
+		});
+	}
 
-	public getProcessableTransactions(
-		limit: number,
-	): ReadonlyArray<Transaction> {}
+	public existsInTransactionPool(transaction: Transaction): boolean {
+		return Object.keys(this._queues).reduce(
+			(previousValue, currentValue) =>
+				previousValue || this._queues[currentValue].exists(transaction),
+			false,
+		);
+	}
 
-	public onDeleteBlock(block: Block): void {}
+	public get queues(): Queues {
+		return this._queues;
+	}
 
-	public onNewBlock(block: Block): void {}
+	public getProcessableTransactions(limit: number): ReadonlyArray<Transaction> {
+		return this._queues.ready.dequeueUntil(
+			queueCheckers.returnTrueUntilLimit(limit),
+		);
+	}
 
-	public onRoundRollback(delegates: ReadonlyArray<string>): void {}
+	public onDeleteBlock(block: Block): void {
```

Comment:
`TransactionPool` as a library should have nothing to do with other entities. It should just expose interfaces to be used based on business logic. And that business logic is not part of the library itself. Otherwise creating a library for transaction-pool does not make sense, everything could be just appear into Lisk Core Module. 

Being library It should expose interfaces like: 

1. moveTransactions(transactions, queue1, queue2)
2. resetTransactions(transactions) 
3. emptyQueue(queue)
4. resetPool

Additionally current approach is making `TransactionPool` to be tightly bounded to `Block` and other structures, which is completely out of its context. `TransactionPool` library should only deal with transactions and queues.

Reply:
Thanks for the review. In the transaction pool library, on certain events in the blockchain,  the transactions are moved from one queue to the other based on the behaviour of different types of transactions described in the specification of the LIP. The behavior of transactions is relevant to transactions in general and doesn't necessarily have to do with transactions in core. For-example, if a new block is received which contains transactions which have unique data, then transactions in the transaction pool which also have unique data should be reverified again, therefore they are moved to a different queue.

The scenario I just described is independent of the business logic of Lisk-core, it is based on how the transactions which require unique data behave in the blockchain. An important thing to note here is that transaction pool does not decide by itself which transaction has unique data, and which don't, it acquires this information by looking at the transaction definition. 

The business logic relevant to lisk-core will be supplied to the library, for-example, the functions for validating, verifying or applying transactions will be written in Core and transaction pool library will use these functions to move the transactions to the relevant queues.

The exposed interface you described looks like a queue library and not a transaction pool library. I think the behavior of transaction pool when a block is added/removed is relevant to core as well as sidechains, therefore, it should be incorporated in the library. 

As for the last point, I somewhat agree with your point, the transaction pool shouldn't care about the block structure, we can update these events, and use `onDeletedTransactions` etc. And hence, the transaction pool library will expect an interface of transactions.

Reply:
Meaning of a library means which can be used by irrespective of the core. If our library can only be used in Lisk Core then there is no need of splitting up transaction pool into module and library. 

So transaction pool library is just provide a Queue interface specialized for transctions. So along with transaction library anyone should be able to use that library in any project. 

So rather than exposing events specified by our blockchain it should just expose interfaces based on behavior. e.g. instead of ‘onDeleteTransaction’ it should expose interface ‘deleteTransactions’ and leave the business logic to implementation. If we want to delete on new block or some other condition. 

And that’s just an example to straighten up the ideology of any library in lisk-elements.

Reply:
In short, instead of exposing event handlers, just expose interfaces. And yes trigger events that are strictly associated to library. e.g. if I want to know when a transaction changed its queu I can listen to an event of transaction pool.

Reply:
I don't think it's better to have
```
moveTransactions(transactions, queue1, queue2)
resetTransactions(transactions)
emptyQueue(queue)
resetPool
```
these interfaces. Since transaction pool should manage it's internal state.
but I agree that instead of implementing handler, it should provide more generic functions.
so instead of `onNewBlock` => `deleteTransactions` sounds good to me